### ISSUE TITLE
Race condition causing hang in vfs.readAt()

### ISSUE THREAD

**Issue by cafarm** [rclone/rclone#4039]
#### What is the problem you are having with rclone?
rclone mount stops responding and hangs indefinitely. Killing rclone when this happens reveals the following stack:

```
goroutine 8157 [sync.Cond.Wait, 79 minutes]:
runtime.goparkunlock(...)
      /usr/local/go/src/runtime/proc.go:310
sync.runtime_notifyListWait(0xc00008e210, 0x16d)
      /usr/local/go/src/runtime/sema.go:510 +0xf8
sync.(*Cond).Wait(0xc00008e200)
      /usr/local/go/src/sync/cond.go:56 +0x8e
github.com/rclone/rclone/vfs.(*ReadFileHandle).readAt(0xc000606000, 0xc0001f6000, 0x2000, 0x2000, 0x298b2a000, 0xc0001f6000, 0x17, 0x4)
      /go/rclone/vfs/read.go:253 +0x1dd
github.com/rclone/rclone/vfs.(*ReadFileHandle).ReadAt(0xc000606000, 0xc0001f6000, 0x2000, 0x2000, 0x298b2a000, 0x0, 0x0, 0x0)
      /go/rclone/vfs/read.go:213 +0xec
github.com/rclone/rclone/cmd/mount.(*FileHandle).Read(0xc00057e240, 0x2a18ce0, 0xc0000aa4c0, 0xc0001222a0, 0xc000682200, 0x1, 0xc000237100)
      /go/rclone/cmd/mount/handle.go:29 +0x3eb
bazil.org/fuse/fs.(*Server).handleRequest(0xc0002e23c0, 0x2a18ce0, 0xc0000aa4c0, 0x29e9f40, 0xc0003198c8, 0xc0000aaac0, 0x2a0fba0, 0xc0001222a0, 0xc000273ec0, 0xc00036ce58, ...)
      /root/go/pkg/mod/bazil.org/fuse@v0.0.0-20191225233854-3a99aca11732/fs/serve.go:1246 +0xe14
bazil.org/fuse/fs.(*Server).serve(0xc0002e23c0, 0x2a0fba0, 0xc0001222a0)
      /root/go/pkg/mod/bazil.org/fuse@v0.0.0-20191225233854-3a99aca11732/fs/serve.go:875 +0x54c
bazil.org/fuse/fs.(*Server).Serve.func1(0xc0002e23c0, 0x2a0fba0, 0xc0001222a0)
      /root/go/pkg/mod/bazil.org/fuse@v0.0.0-20191225233854-3a99aca11732/fs/serve.go:424 +0x8a
created by bazil.org/fuse/fs.(*Server).Serve
      /root/go/pkg/mod/bazil.org/fuse@v0.0.0-20191225233854-3a99aca11732/fs/serve.go:422 +0x602
 
goroutine 8112 [select, 79 minutes]:
github.com/rclone/rclone/fs/asyncreader.(*AsyncReader).init.func1(0xc0000e4000)
      /go/rclone/fs/asyncreader/asyncreader.go:83 +0x1e2
created by github.com/rclone/rclone/fs/asyncreader.(*AsyncReader).init
      /go/rclone/fs/asyncreader/asyncreader.go:78 +0x293
```

I believe this is because there is a race condition in the following section of code:

https://github.com/rclone/rclone/blob/5ee24f804f0058d33da4ebee2235a53508f770e5/vfs/read.go#L238-L255

If `timeout` fires before it gets down to `fh.cond.Wait()` on line 254, I believe it will be stuck waiting for a broadcast that will never occur.

#### What is your rclone version (output from `rclone version`)

```
rclone v1.51.0
- os/arch: linux/amd64
- go version: go1.13.7
```

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

Ubuntu 16.04 64 bit

####  Which cloud storage system are you using? (eg Google Drive)

Google Cloud Storage

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

`rclone mount my-container: myfolder/`

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

N/A

**Comment by ncw** [comment]
Hmm, have you got the full backtrace?

I'd like to see if the go routine started at line 242 was still running.

How did you get this to happen? Does it happen often?

> If `timeout` fires before it gets down to `fh.cond.Wait()` on line 254, I believe it will be stuck waiting for a broadcast that will never occur.

What you are saying I think is that the `fh.cond.Broadcast()` on line 248 runs before the `fh.cond.Wait() ` on line 254 and hence the `Wait` waits forever.

I'll just note that we set `atomic.StoreInt32(&abort, 1) ` on line 246 which is tested at line 252.

I think this makes the race window quite small, but I think it is still there.

Any ideas for a fix? Unfortunately go doesn't have a Cond lock with timeout otherwise this would be a lot easier...

I've re-written this code already as a message sending code which makes it much easier to reason about, but I'm not sure about the performance.

**Comment by cafarm** [comment]
The full backtrace is below.

> How did you get this to happen? Does it happen often?

It occurs infrequently (probably less than a percent of a percent). We can fairly reliably get it to occur by looping over an `rclone mount` sequence a few hundred times. That sequence will mount a remote, access some data for about a minute, unmount and repeat.

> I'll just note that we set atomic.StoreInt32(&abort, 1) on line 246 which is tested at line 252.
> 
> I think this makes the race window quite small, but I think it is still there.

Indeed. I think the main goroutine would need to make it to the `waiting for in-sequence...` debug statement before `abort` is set to 1. Then the timeout goroutine would need to set `abort` to 1 and broadcast before the main routine hits wait. It is definitely a small window but the frequency of occurrence seems to indicate the issue would be caused by something with a small window.

> Any ideas for a fix?

Perhaps using the `cond` Locker to make sure wait is called before broadcast?

```go
maxWait := fh.file.d.vfs.Opt.ReadWait
timeout := time.NewTimer(maxWait)
done := make(chan struct{})
abort := int32(0)
go func() {
	select {
	case <-timeout.C:
		fh.cond.L.Lock()
		// set abort flag an give all the waiting goroutines a kick on timeout
		atomic.StoreInt32(&abort, 1)
		fs.Debugf(fh.remote, "aborting in-sequence read wait, off=%d", off)
		fh.cond.Broadcast()
		fh.cond.L.Unlock()
	case <-done:
	}
}()
for fh.offset != off && atomic.LoadInt32(&abort) == 0 {
	fs.Debugf(fh.remote, "waiting for in-sequence read to %d for %v", off, maxWait)
	fh.cond.Wait()
}
fh.cond.L.Unlock()
```

This assumes `fh.cond.L` is locked before coming in here. I'm not sure if that's a good assumption? 

```
goroutine 0 [idle]:
runtime.futex(0x3784e68, 0x80, 0x0, 0x0, 0xc000000000, 0xb251b7, 0x3784700, 0xb0b08d, 0x7ffd10ecb278, 0xb0b3af, ...)
      /usr/local/go/src/runtime/sys_linux_amd64.s:535 +0x21
runtime.futexsleep(0x3784e68, 0x100000000, 0xffffffffffffffff)
      /usr/local/go/src/runtime/os_linux.go:44 +0x46
runtime.notesleep(0x3784e68)
      /usr/local/go/src/runtime/lock_futex.go:151 +0x9f
runtime.stopm()
      /usr/local/go/src/runtime/proc.go:1928 +0xc0
runtime.findrunnable(0xc00004aa00, 0x0)
      /usr/local/go/src/runtime/proc.go:2391 +0x53f
runtime.schedule()
      /usr/local/go/src/runtime/proc.go:2524 +0x2be
runtime.park_m(0xc00009ec00)
      /usr/local/go/src/runtime/proc.go:2610 +0x9d
runtime.mcall(0x213c5c0)
      /usr/local/go/src/runtime/asm_amd64.s:318 +0x5b
 
goroutine 1 [select, 81 minutes]:
github.com/rclone/rclone/cmd/mount.Mount(0x2a323c0, 0xc00011e780, 0x7ffd10ecbdf4, 0x8, 0x2592a88, 0x1)
      /go/rclone/cmd/mount/mount.go:150 +0x3db
github.com/rclone/rclone/cmd/mountlib.NewMountCommand.func1(0xc000484f00, 0xc000578210, 0x2, 0xb)
      /go/rclone/cmd/mountlib/mount.go:295 +0x3e0
github.com/spf13/cobra.(*Command).execute(0xc000484f00, 0xc0000f8b00, 0xb, 0xb, 0xc000484f00, 0xc0000f8b00)
      /root/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:830 +0x8e1
github.com/spf13/cobra.(*Command).ExecuteC(0x36fe460, 0x5e616adb, 0x3783a60, 0xc0000ac010)
      /root/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x41b
github.com/spf13/cobra.(*Command).Execute(...)
      /root/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
github.com/rclone/rclone/cmd.Main()
      /go/rclone/cmd/cmd.go:506 +0x105
main.main()
      /go/rclone/rclone.go:14 +0x30
 
goroutine 19 [syscall, 81 minutes]:
os/signal.signal_recv(0xb623b1)
      /usr/local/go/src/runtime/sigqueue.go:147 +0x9c
os/signal.loop()
      /usr/local/go/src/os/signal/signal_unix.go:23 +0x30
created by os/signal.init.0
      /usr/local/go/src/os/signal/signal_unix.go:29 +0x4f
 
goroutine 22 [select]:
go.opencensus.io/stats/view.(*worker).start(0xc0000deeb0)
      /root/go/pkg/mod/go.opencensus.io@v0.22.2/stats/view/worker.go:154 +0x1d6
created by go.opencensus.io/stats/view.init.0
      /root/go/pkg/mod/go.opencensus.io@v0.22.2/stats/view/worker.go:32 +0x9a
 
goroutine 51 [syscall, 79 minutes]:
syscall.Syscall(0x0, 0x9, 0xc000892000, 0x21000, 0xc00075a180, 0x21000, 0xc000892000)
      /usr/local/go/src/syscall/asm_linux_amd64.s:18 +0x5
syscall.read(0x9, 0xc000892000, 0x21000, 0x21000, 0xc000138310, 0xc0005dad50, 0xb774ca)
      /usr/local/go/src/syscall/zsyscall_linux_amd64.go:732 +0x68
syscall.Read(0x9, 0xc000892000, 0x21000, 0x21000, 0x50, 0x0, 0x0)
      /usr/local/go/src/syscall/syscall_unix.go:183 +0x5b
bazil.org/fuse.(*Conn).ReadRequest(0xc0000927e0, 0x2628810, 0xc0002e23c0, 0x2a0fba0, 0xc000122310)
      /root/go/pkg/mod/bazil.org/fuse@v0.0.0-20191225233854-3a99aca11732/fuse.go:558 +0x147
bazil.org/fuse/fs.(*Server).Serve(0xc0002e23c0, 0x29e9f20, 0xc0000c4260, 0x0, 0x0)
      /root/go/pkg/mod/bazil.org/fuse@v0.0.0-20191225233854-3a99aca11732/fs/serve.go:413 +0x627
github.com/rclone/rclone/cmd/mount.mount.func1(0xc0002e23c0, 0xc0000c4260, 0xc0000927e0, 0xc000138360)
      /go/rclone/cmd/mount/mount.go:96 +0x53
created by github.com/rclone/rclone/cmd/mount.mount
      /go/rclone/cmd/mount/mount.go:95 +0x443
 
goroutine 68 [select]:
github.com/rclone/rclone/fs/accounting.(*Account).averageLoop(0xc0000bb790)
      /go/rclone/fs/accounting/accounting.go:146 +0x1d7
created by github.com/rclone/rclone/fs/accounting.newAccountSizeName
      /go/rclone/fs/accounting/accounting.go:66 +0x3a7
 
goroutine 8157 [sync.Cond.Wait, 79 minutes]:
runtime.goparkunlock(...)
      /usr/local/go/src/runtime/proc.go:310
sync.runtime_notifyListWait(0xc00008e210, 0x16d)
      /usr/local/go/src/runtime/sema.go:510 +0xf8
sync.(*Cond).Wait(0xc00008e200)
      /usr/local/go/src/sync/cond.go:56 +0x8e
github.com/rclone/rclone/vfs.(*ReadFileHandle).readAt(0xc000606000, 0xc0001f6000, 0x2000, 0x2000, 0x298b2a000, 0xc0001f6000, 0x17, 0x4)
      /go/rclone/vfs/read.go:253 +0x1dd
github.com/rclone/rclone/vfs.(*ReadFileHandle).ReadAt(0xc000606000, 0xc0001f6000, 0x2000, 0x2000, 0x298b2a000, 0x0, 0x0, 0x0)
      /go/rclone/vfs/read.go:213 +0xec
github.com/rclone/rclone/cmd/mount.(*FileHandle).Read(0xc00057e240, 0x2a18ce0, 0xc0000aa4c0, 0xc0001222a0, 0xc000682200, 0x1, 0xc000237100)
      /go/rclone/cmd/mount/handle.go:29 +0x3eb
bazil.org/fuse/fs.(*Server).handleRequest(0xc0002e23c0, 0x2a18ce0, 0xc0000aa4c0, 0x29e9f40, 0xc0003198c8, 0xc0000aaac0, 0x2a0fba0, 0xc0001222a0, 0xc000273ec0, 0xc00036ce58, ...)
      /root/go/pkg/mod/bazil.org/fuse@v0.0.0-20191225233854-3a99aca11732/fs/serve.go:1246 +0xe14
bazil.org/fuse/fs.(*Server).serve(0xc0002e23c0, 0x2a0fba0, 0xc0001222a0)
      /root/go/pkg/mod/bazil.org/fuse@v0.0.0-20191225233854-3a99aca11732/fs/serve.go:875 +0x54c
bazil.org/fuse/fs.(*Server).Serve.func1(0xc0002e23c0, 0x2a0fba0, 0xc0001222a0)
      /root/go/pkg/mod/bazil.org/fuse@v0.0.0-20191225233854-3a99aca11732/fs/serve.go:424 +0x8a
created by bazil.org/fuse/fs.(*Server).Serve
      /root/go/pkg/mod/bazil.org/fuse@v0.0.0-20191225233854-3a99aca11732/fs/serve.go:422 +0x602
 
goroutine 8112 [select, 79 minutes]:
github.com/rclone/rclone/fs/asyncreader.(*AsyncReader).init.func1(0xc0000e4000)
      /go/rclone/fs/asyncreader/asyncreader.go:83 +0x1e2
created by github.com/rclone/rclone/fs/asyncreader.(*AsyncReader).init
      /go/rclone/fs/asyncreader/asyncreader.go:78 +0x293
```

**Comment by Rootax** [comment]
Even if I have no way to help, i hope this problem is not forgotten : )

**Comment by Rootax** [comment]
Thx for the commit :)

**Comment by ncw** [comment]
I managed to replicate this locally

In one terminal

```
mkdir /tmp/src
truncate -s 1GB /tmp/src/1GB
rclone mount -vv /tmp/src /mnt/tmp/ --vfs-read-wait 0.1ms
```

And in another

```
for i in `seq 100` ; do echo $i ; cat /mnt/tmp/1GB >/dev/null ;done
```

This locked up after 7,5,19,15,24 iterations in 5 tests so an average of 14 tests....

I ran this with the fix 300 times without failure for a probability of [2E-277](https://www.wolframalpha.com/input/?i=1-Q%28300%2C14%29) of this being a fluke (everyone loves a poisson distrubution ;-)

I also ran it with the race detector for 20 iterations (it was veeeery slow).

I've merged this fix to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.50





I was hoping to fix this code with a reinvention of how it works but I haven't finished doing that yet...
